### PR TITLE
Ray-level OOM killer

### PR DESCRIPTION
**Pull request by clarng**
## Why are these changes needed?

The Ray-level OOM killer preemptively evicts worker when the node is under memory pressure. This PR adds a memory monitor that runs periodically to check if the node memory usage is above a certain threshold. The caller may provide a callback to the monitor to execute at each interval to determine whether an action should be taken.

The node manager implements the callback to evict the worker with the newest task. It evicts only one worker at a time and enforces that by tracking the last evicted worker. If the eviction is still in progress it will not evict another worker even if the usage is above the threshold.

This PR is a no-op since the monitor is disabled by default.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by wuisawesome** [comment]
Potentially dumb question, but what's the difference between this and `memory_monitor.py`? https://github.com/ray-project/ray/blob/master/python/ray/_private/memory_monitor.py

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -331,7 +331,51 @@ NodeManager::NodeManager(instrumented_io_context &io_service,
       record_metrics_period_ms_(config.record_metrics_period_ms),
       next_resource_seq_no_(0),
       ray_syncer_(io_service_, self_node_id_.Binary()),
-      ray_syncer_service_(ray_syncer_) {
+      ray_syncer_service_(ray_syncer_),
+      memory_monitor_(std::make_unique<MemoryMonitor>(
```
Comment (scv119) [inline src/ray/raylet/node_manager.cc:335]: i wonder if we can put this code into helper function?

File: python/ray/data/context.py [hunk 2]
```diff
@@ -163,6 +163,7 @@ def get_current() -> "DatasetContext":
                 _default_context.block_splitting_enabled
                 and _default_context.block_owner is None
             ):
+                print("block_splitting_enabled true")
```
Comment (scv119) [inline python/ray/data/context.py:166]: remove this

**Review by scv119 — COMMENTED** [review]
first round of review. The code looks very clean! Need to clean up the python side a little bit.

**Comment by scv119** [comment]
> Potentially dumb question, but what's the difference between this and memory_monitor.py? https://github.com/ray-project/ray/blob/master/python/ray/_private/memory_monitor.py

pretty much the same but we need to do it in core so it's easier to interact with scheduler.

**Comment by clarng** [comment]
> Potentially dumb question, but what's the difference between this and `memory_monitor.py`? https://github.com/ray-project/ray/blob/master/python/ray/_private/memory_monitor.py

The current class is just a checker used at submission time and not a monitor, and is in python and doesn't have the information needed to react to memory pressure

**Review by rkooo567 — COMMENTED** [review]
Really excited about the PR! Will review it soon! Btw, does it make sense to split the PR to some smaller multiple PRs? Maybe we can split it to (1) memory monitor only and (2) integration to node manager (3) turn on?

**Comment by clarng** [comment]
> Really excited about the PR! Will review it soon! Btw, does it make sense to split the PR to some smaller multiple PRs? Maybe we can split it to (1) memory monitor only and (2) integration to node manager (3) turn on?

Broke out the memory monitor - https://github.com/ray-project/ray/pull/27017

File: python/ray/data/context.py (hunk 2 above)
Comment (clarng) [reply python/ray/data/context.py]: this was pushed by accident, reverted

File: python/ray/_private/serialization.py [hunk 3]
```diff
@@ -281,6 +281,10 @@ def _deserialize_object(self, data, metadata, object_ref):
                 return OutOfDiskError(
                     object_ref.hex(), object_ref.owner_address(), object_ref.call_site()
                 )
+            elif error_type == ErrorType.Value("OUT_OF_MEMORY_ERROR"):
+                return OutOfDiskError(
```
[the remark this answers is not in the document]
Comment (clarng) [reply python/ray/_private/serialization.py]: this was pushed by accident, reverted

File: python/ray/data/_internal/compute.py [hunk 4]
```diff
@@ -450,6 +458,22 @@ def _map_block_nosplit(
         builder.add_block(new_block)
     new_block = builder.build()
     accessor = BlockAccessor.for_block(new_block)
+
+    end_uss = int(process.memory_full_info().uss)
+    end_rss = int(process.memory_full_info().rss)
+    uss_delta = end_uss - start_uss
+    available = psutil.virtual_memory().available
+    print(f'delta {uss_delta} end {end_uss} end-rss {end_rss} av {available} pid {os.getpid()}')
```
[the remark this answers is not in the document]
Comment (clarng) [reply python/ray/data/_internal/compute.py]: this was pushed by accident, reverted

File: ssml_cuj_bp.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
+import traceback
```
[the remark this answers is not in the document]
Comment (clarng) [reply ssml_cuj_bp.py]: this was pushed by accident, reverted

File: src/ray/raylet/node_manager.cc (hunk 1 above)
Comment (clarng) [reply src/ray/raylet/node_manager.cc:335]: sg

**Comment by clarng** [comment]
Superseded by https://github.com/ray-project/ray/pull/27384

**Pull request closed without merging**